### ISSUE TITLE
Customizer still running after finished and closed.

### ISSUE THREAD

**Issue by ghost** [clearkimura/Customizer#14]
I used Customizer to create my own Ubuntu Mini Remix version. After finished I clicked "clean", which seemed to work fine, and closed the app. The window did close.

Some hours later I was checking system tasks for another reason and saw that Customizer was still running with some subprocesses and appeared to still have root access. I find this disconcerting and was immediately suspicious of intent. I killed them manually but there must be some problem here.

Please contact me if you need more info to track this down.

**Comment by fluxer** [comment]
That's strange, really strange. What process was running -- the Bash script (customizer) or the GUI (GUI.gambas). If the later then probably that's an issue with Gambas it self.

You can always take a look at the sources and see that there isn't a singe malicous line of code but if you think there is something suspicious I will explain the code to you (if you point to the part you are interested in) and provide a link to the Gambas Documentation with more information so you know that Customizers intentions are just to help you out and not run as deamon in the background and do something it's not suposed to.

Cheers!

**Comment by ghost** [comment]
I couldn't say for sure now. There was nothing on the desktop so maybe it was a bash script. I recall it had something in all caps like ROOT GAMBAS or something. I'm fairly sure it wasn't malicious but when I see stuff running I thought was finished my first thought is to be suspicious of it.

It looks like a good program but in the end my final boot image didn't boot. It did work in VirtualBox and QEMU but physical boot just went dead. I'll have to try again later when I have more time. It was very slow (due to squashfs I think). It would be nice if there was an option that allowed creating a non-squashfs boot image.

**Comment by fluxer** [comment]
The slow boot is a problem that is know in the recent releases of Ubuntu -- 11.10 is one that I've tried and experienced that, it's because Ubuntu is trying to configure your internet (DHCP) and couldn't do it (for whatever the reason is). '

If you think that squashfs is the problem you can try different compression -- gzip instead of xz which is the default for Customizer but my tests show that is slows down the I/O (xz). And by non-squashfs image what do you mean? ext3, ext4, etc.? If that is known to work with the default kernels of Ubuntu (it should) I can do that (in the dev version which has some features still unfinished and isn't going to be finished soon), if you can test images and such as I'm no longer running Ubuntu nor am planning to. 

Oh, and BTW if you report your issues there is a better chance for them to be fixed,like this one with the ISO boot problem (if it is Customizers fault), right? So, please do and remember to provide as much information as possible.

**Comment by ghost** [comment]
@fluxer 
For any submitted issue, users shall provide at least:
- what version of system host (where customizer is installed i.e. 10.04, 12.04)?
- what version of customizer (i.e. 3.2.1)?
- what ISO image (i.e. ubuntu-mini-remix-12.04-amd64)?
- description of problem + screenshot (if possible)

@bkkcoins 
Kindly provide above details for this issue.
So I could test on my machine and check if this issue can be reproduced.

**Comment by ghost** [comment]
Ubuntu 11.04
Customizer from repo, 3.2.2-20120719~109~natty1
ubuntu-mini-remix-12.04-i386

Finished build and clicked clean and closed window. 
Several hours later I found customizer still running.
(shown in htop, looked like bash script with ROOT in cmdline, had some sub processes I believe)

I have removed customizer from my system now and not attempted to re-create.

**Comment by fluxer** [comment]
@bkkcoins Wow, do you really think it's dangerous using Customizer? I really hope we can clear this out.

@clearkimura Thanks for the input, mate! I hope you have the time to do some tests.

**Comment by ghost** [comment]
@bkkcoins 
I've tested using the same Customizer from PPA repo, 3.2.2-20120719~109 and ISO file of ubuntu-mini-remix-12.04-i386 like you had used.

Result: I failed to reproduce error as reported by you.

Customizer runs processes with these keywords: `gksu` `root` `gambas`
Say, if your current session was ended by force [while you are running Customizer] and you had to log in again, these processes will continue to run in background even Customizer window does not appear on your Desktop. At least, this is what I had experienced before.

In other words, what you had experienced is _not_ a bug. That is due to inappropriate handling of session management--or the session was closed abruptly.

Have I solved your issue?

**Comment by ghost** [comment]
I don't have an issue because I've stopped using the program. But if you're asking if I somehow didn't notice that I was logged out while building an image then I can state definitively that I was not. It is possible that canceling a build didn't shutdown a background process as I may have tried to stop a build part way. Does your program capture signals to ensure that subprocesses get cleanly closed if the program is terminated?

**Comment by ghost** [comment]
@bkkcoins 
To be exact, when running Customizer, three dependant processes will be present: `sudo` `gksu` `su-to-root`. In this case, you can check with command `ps -A | grep su` in Terminal.

Now let's go to test cases from start to end:

Case [1]
When I clicked on "Select ISO", chroot Terminal window appears, begin to load the image file and shows progress meter [===\   ]. While still loading, I closed it using window control button (x). The window closes and a warning dialog is displayed, request to start over again. Click "OK", "Clean", close main window.

> Result:   No root process running in background.

Case [2]
After successfully loads an ISO file, click "Terminal" button on right. The chroot Terminal will appear. Wait until it's ready to type command. Try closing the main window using window control button (x). It will not close, since main window does not respond while chroot Terminal opens. But when you exit chroot Terminal, previously clicked action is executed so the main window will follow to close immediately.

> Result:   No root process running in background.

Case [3]
After successfully installs needed packages to build minimal system, click "Rebuild ISO". The chroot Terminal will appear. Wait until it starts to show the progress meter [===\      ]. Try closing the Terminal using window control button (x). It closes immediately. Then close the main window as usual.

> Result:   No root process running in background.

I tried all three cases above on first hand, attempting to disturb the processes by closing chroot Terminal abruptly, yet not even once root process failed to terminate.

To explain any possibility that might happen, we shall hear from the developer.

@fluxer 
Any technical details i.e. capture signal issue to be explained, if you may?

**Comment by fluxer** [comment]
Signals are not trapped, logical 'or' statements are used to catch errors in place. They handle keyboard interupts (ctrl + c) and exit status different than 0 but first output an error message and until a key is pressed the script remains running. That may be the case (a Terminal was left on one of your virtual desktops/workspaces with the application running).

Anyway, unless you wish to test it again and make sure that you can reproduce the issue (or someone else out there) I am going to close the issue after certain time of inactivity on your side, we need more information. Me and clearkimura cannot reproduce this.
